### PR TITLE
New feature: Total Variation (image_ops)

### PR DESCRIPTION
**Pull request by Hvass-Labs**
This is a pull-request for the new feature discussed in https://github.com/tensorflow/tensorflow/issues/6397

I have made a short Python Notebook for testing this:

https://gist.github.com/Hvass-Labs/ac4ef0074fd182c6778532ba222d8c37

I can add a unit-test when the API / semantics have been agreed upon. I suppose I have to build TensorFlow locally in order to run a unit-test? Have you made a guide for doing this? I have some horrible experiences trying to build other open-source projects in the past.

----

PS: You may want to have your legal department go over the CLA agreement. It is legalistic gibberish and I had to read parts of it several times to sort-of understand it. Example: "For legal entities, the entity making a Contribution and all other entities that control, are controlled by, or are under common control with that entity are considered to be a single Contributor. For the purposes of this definition, "control" means (i) the power, direct or indirect, to cause the direction or management of such entity, whether by contract or otherwise, or (ii) ownership of fifty percent (50%) or more of the outstanding shares, or (iii) beneficial ownership of such entity."

### CODE REVIEW

**Comment by shlens** [comment]
The API and semantics look good. Please do add a unit test before we begin reviewing the code. Thank you.

We do have a guide for building from source. Please take a look here:
https://www.tensorflow.org/get_started/os_setup#installing_from_sources

**Comment by Hvass-Labs** [comment]
I have now added some unit-tests to the pull-request. I have tried to follow the style in `image_ops_test.py`

I have a few remarks:

- It appears that some of the tests in `image_ops_test.py` merely test the TensorFlow implementation against an equivalent numpy implementation. I'm not sure I agree that this is a good way of testing, because you're likely to make the same mistakes in both implementations. Nevertheless, I didn't want to handcraft large test-cases, so I did the same :-) This has been supplemented with a few small hand-made test-cases, and I also tested the function using a Python Notebook - which is a superior kind of testing, but it cannot be automated.

- I couldn't quite figure out the naming conventions used for the private test-functions but it is presumably of minor importance how I named private functions in these unit-tests. You may rename them if you wish.

- I also wanted to test if the proper exception was raised for incorrect arguments, but `assertRaises()` didn't really seem suited for that purpose.

- As usual, my code-comments are more verbose. I believe this is very helpful to those of us who don't work on TensorFlow-core every day, as it makes it easier for us to grasp what the intention is with the code.

- I have not tested using bazel. Officially bazel only supports Ubuntu 15 and I use Ubuntu 16. I did some internet searches and found out that bazel installation may involve several steps and dependencies, and it could be difficult or impossible to remove if something went wrong. I didn't want to take the risk of breaking my Linux or TensorFlow installation just to run this unit-test, as it would take me several days to reinstall everything. So I ran the test-code in the same Python Notebook provided above, and just hacked it slightly with a few imports so it could run there. It should run in bazel. Please test this.

**Comment by shlens** [comment]
Jenkins, test this please.

**Comment by shlens** [comment]
The test fails:
`Traceback (most recent call last):
  File "/workspace/pip_test/tests/image_ops_test.py", line 2431, in testTotalVariationHandmade
    self._test(a, tot_var)
  File "/workspace/pip_test/tests/image_ops_test.py", line 2301, in _test
    y = image_ops.total_variation(images=x_tf)
AttributeError: 'module' object has no attribute 'total_variation`

**Comment by vrv** [comment]
To make the symbol available in the image_ops namespace, I think you'll have to add the @@total_variation  callout to image_ops.py: https://github.com/Hvass-Labs/tensorflow/blob/02da06577be86ed095cde20fc674e7604e99894d/tensorflow/python/ops/image_ops.py#L164

**Comment by vrv** [comment]
cc: @martinwicke @gunan so they know this is an API being introduced

**Comment by martinwicke** [comment]
thanks @vrv, sgtm

**Comment by Hvass-Labs** [comment]
Thanks for your help and patience.

I have now tried to export the function in the image_ops namespace, as you said.

I'm still reluctant to install bazel and build tensorflow from source in case it breaks my working installation, because I'm currently not planning on making a lot of contributions to tf-core. So I'll see if I can get Jenkins to test it as well - I'm assuming it is a bot:

Jenkins, test this please.

(Reminds me of this :-) https://www.youtube.com/watch?v=MFqxdvggAxM

**Comment by gunan** [comment]
Jenkins, test this please.

File: tensorflow/python/ops/image_ops_impl.py [hunk 1]
```diff
@@ -1259,3 +1259,73 @@ def _jpeg():
 
     is_jpeg = math_ops.equal(substr, b'\xff\xd8\xff\xe0', name='is_jpeg')
     return control_flow_ops.cond(is_jpeg, _jpeg, check_png, name='cond_jpeg')
+
+
+def total_variation(images, name=None):
+  """Calculate and return the Total Variation for one or more images.
```
Comment (shlens) [inline tensorflow/python/ops/image_ops_impl.py]: Uncapitalize "Total Variation" since it is not a proper name?

**Review by shlens — APPROVED** [review]
Please fix minor comment but otherwise the change looks great.

**Comment by Hvass-Labs** [comment]
I changed the capitalization of Total Variation in the comment, as you requested. I prefer it in capitals so it stands out, but I won't argue :-)

Thanks for the help and I'm glad you like the contribution!

**Comment by shlens** [comment]
Jenkins, test this please.

**Comment by shlens** [comment]
Jenkins, test this please.

**Comment by yaroslavvb** [comment]
BTW Bazel takes me about 2 mins to install using following commands:

```
sudo add-apt-repository ppa:webupd8team/java
sudo apt-get update
sudo apt-get install oracle-java8-installer

echo "deb [arch=amd64] http://storage.googleapis.com/bazel-apt stable jdk1.8" | sudo tee /etc/apt/sources.list.d/bazel.list
curl https://bazel.build/bazel-release.pub.gpg | sudo apt-key add -
sudo apt-get update && sudo apt-get install bazel
sudo apt-get upgrade bazel
sudo apt-get upgrade bazel --fix-missing

```

**Comment by gunan** [comment]
Also, you can always use docker to create an isolated container which should not affect your working setup at all.
You can always download devel docker images and work with them.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by Hvass-Labs** [comment]
It seems that two of the tests are failing. I looked at the error message which was VERY long. I searched for `total_variation` but it does not appear in the errors, so I don't know if the errors are related to my contributed code? Also, the other tests seem fine. Should I do anything?

Thanks for the tips on installing bazel. My concern was both whether it might break Linux - and if it would somehow override my working installations of TensorFlow. If I'm only making one or two contribs to tf-core then it didn't seem worth the risk. But I've noted your installation instructions in case I'll need to build tf in the future.

**Comment by drpngx** [comment]
The error is this:

```
(482 / 518) Python test-on-install FAILED (32988 ms): tensorflow/python/training/basic_session_run_hooks_test.py
  Log @: /workspace/pip_test/tests/logs/tensorflow/python/training/basic_session_run_hooks_test.py.log
============== BEGINS failure log content ==============
....W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
I tensorflow/compiler/xla/service/platform_util.cc:58] platform Host present with 32 visible devices
I tensorflow/compiler/xla/service/service.cc:180] XLA service executing computations on platform Host. Devices:
I tensorflow/compiler/xla/service/service.cc:187]   StreamExecutor device (0): <undefined>, <undefined>
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
FW tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
..W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
..W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
..W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
..W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
........W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
..W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
...W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.....W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
.W tensorflow/compiler/jit/xla_gpu_device.cc:48] Failed to create XLA_GPU device: Not found: could not find registered platform with name: "CUDA"
..
======================================================================
FAIL: test_save_secs_saves_periodically (__main__.CheckpointSaverHookTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/pip_test/tests/basic_session_run_hooks_test.py", line 343, in test_save_secs_saves_periodically
    self.global_step.name))
AssertionError: 3 != 4

----------------------------------------------------------------------
Ran 47 tests in 30.475s

FAILED (failures=1)
============== ENDS failure log content ==============
```

**Comment by drpngx** [comment]
@gunan @hawkinsp it's running XLA in Linux CPU Tests?

**Comment by hawkinsp** [comment]
@drpngx XLA is linked in at the moment, but it should not actually be doing anything in this test. The warning is benign and comes from device initialization. It should probably be lowered in severity; it just means you don't have a GPU, which this builder does not.

I think that test is simply flaky (it uses sleep() and expects a certain number of events to occur in a certain time). Perhaps XLA made something a touch slower, but perhaps not.

Either way I think @gunan is working on disabling XLA in the regular builds and adding a separate XLA-only build, which should fix any problems.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
144150966 is supposed to fix XLA linking in all builds. So builds should recover once that is pushed to github.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Pull request merged**